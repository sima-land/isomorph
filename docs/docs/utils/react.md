---
title: Утилиты для React
description: Утилиты для работы с React.
---
# Утилиты для React

Пакет предоставляет утилиты для работы с React.

### ErrorBoundary

- Компонент перехватчик ошибок, обрабатывает все ошибки при рендере дочерних элементов.
- Рендерит fallback если была выявлена ошибка при рендере всех дочерних элементов.

##### Пример использования:

```tsx
import React from 'react';
import { App } from './app';
import { ErrorBoundary } from '@sima-land/isomorph/utils/react';

const rootElement = document.getElementById(config.appName);
const onError = logger.error;

createRoot(rootElement).render(
    <ErrorBoundary onError={onError} fallback={null}>
        <App />
    </ErrorBoundary>,
);
```

### SafeSuspense

Компонент обёртки Suspense из React, дополнительно использующая ErrorBoundary в реализации.

##### Пример использования:

```tsx
import React from 'react';
import { App } from './app';
import { SafeSuspense } from '@sima-land/isomorph/utils/react';

const rootElement = document.getElementById(config.appName);
const onError = logger.error;

createRoot(rootElement).render(
    <SafeSuspense onError={onError} fallback={null}>
        <App />
    </SafeSuspense>,
);
```
