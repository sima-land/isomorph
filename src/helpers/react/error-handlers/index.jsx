import React from 'react';

/**
 * Обработчик ошибок в React компонентах.
 */
export class ErrorBoundary extends React.Component {
  /**
   * @param {Object} props Свойства компонента.
   * @param {import('react').ReactNode} props.children Дочерний компонент.
   * @param {import('react').ReactNode} props.fallback Запасной элемент, если возникла ошибка.
   * @param {Function} [props.captureException] Функция для логирования ошибки.
   */
  constructor (props) {
    super(props);
    this.state = { hasError: false };
  }

  /** @inheritdoc */
  static getDerivedStateFromError () {
    return { hasError: true };
  }

  /** @inheritdoc */
  componentDidCatch (error, errorInfo) {
    const { captureException } = this.props;
    captureException?.(error, errorInfo);
  }

  /**
   * Рендер.
   * @return {ReactElement} Дочерний элемент либо запасное значение.
   */
  render () {
    const {
      children,
      fallback,
    } = this.props;

    return this.state.hasError ? fallback : children;
  }
}

/**
 * Обертка над Suspense c перехватом ошибок.
 * @param {Object} props Свойства компонента.
 * @param {import('react').ReactNode} props.children Дочерний компонент.
 * @param {import('react').ReactNode} props.fallback Запасной элемент, если возникла ошибка.
 * @param {Function} [props.captureException] Функция для логирования ошибки.
 * @return {ReactElement} Элемент.
 */
export const SafeSuspense = ({
  children,
  fallback,
  captureException,
}) => (
  <React.Suspense fallback={fallback}>
    <ErrorBoundary
      children={children}
      fallback={fallback}
      captureException={captureException}
    />
  </React.Suspense>
);
