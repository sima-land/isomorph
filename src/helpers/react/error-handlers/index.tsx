import React from 'react';

export interface Props {
  children: NonNullable<React.ReactNode> | null
  fallback: NonNullable<React.ReactNode> | null
  captureException?: (error: Error, errorInfo: React.ErrorInfo) => void
}

/**
 * Обработчик ошибок в React компонентах.
 */
export class ErrorBoundary extends React.Component<Props, { hasError: boolean }> {
  /**
   * @param props Свойства компонента.
   * @param props.children Дочерний компонент.
   * @param props.fallback Запасной элемент, если возникла ошибка.
   * @param props.captureException Функция для логирования ошибки.
   */
  constructor (props: Props) {
    super(props);

    this.state = { hasError: false };
  }

  /** @inheritdoc */
  static getDerivedStateFromError () {
    return { hasError: true };
  }

  /** @inheritdoc */
  componentDidCatch (error: Error, errorInfo: React.ErrorInfo) {
    const { captureException } = this.props;

    captureException?.(error, errorInfo);
  }

  /**
   * Рендер.
   * @return Дочерний элемент либо запасное значение.
   */
  render () {
    const { children, fallback } = this.props;

    return this.state.hasError ? fallback : children;
  }
}

/**
 * Обертка над Suspense c перехватом ошибок.
 * @param props Свойства компонента.
 * @param props.children Дочерний компонент.
 * @param props.fallback Запасной элемент, если возникла ошибка.
 * @param props.captureException Функция для логирования ошибки.
 * @return Элемент.
 */
export const SafeSuspense = ({
  children,
  fallback,
  captureException,
}: Props) => (
  <React.Suspense fallback={fallback}>
    <ErrorBoundary
      children={children}
      fallback={fallback}
      captureException={captureException}
    />
  </React.Suspense>
);
