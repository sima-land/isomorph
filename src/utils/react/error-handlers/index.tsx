import { Component, ErrorInfo, ReactNode, Suspense } from 'react';

/**
 * Опции компонентов ErrorBoundary и SafeSuspense.
 */
export interface Props {
  /** Дочерний элемент. */
  children: NonNullable<ReactNode> | null;

  /** Содержимое, которое будет выведено если возникла ошибка. */
  fallback: NonNullable<ReactNode> | null;

  /** Функция для логирования ошибок. */
  onError?: (error: Error, errorInfo: ErrorInfo) => void;
}

/**
 * Компонент-предохранитель. Обработчик ошибок в React-компонентах.
 */
export class ErrorBoundary extends Component<Props, { hasError: boolean }> {
  /**
   * @param props Свойства.
   */
  constructor(props: Props) {
    super(props);

    this.state = { hasError: false };
  }

  /** @inheritDoc */
  static getDerivedStateFromError() {
    return { hasError: true };
  }

  /** @inheritDoc */
  componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    // @todo добавить возможность брать onError из специального контекста (это упростит передачу из di-контейнера)
    const { onError } = this.props;

    onError?.(error, errorInfo);
  }

  /**
   * Рендер.
   * @return Дочерний элемент либо запасное значение.
   */
  render() {
    const { children, fallback } = this.props;

    return <>{this.state.hasError ? fallback : children}</>;
  }
}

/**
 * Обертка над Suspense c перехватом ошибок.
 * @todo Возможно этот компонент не нужен...
 * @param props Свойства.
 * @return Элемент.
 */
export function SafeSuspense({ children, fallback, onError }: Props) {
  return (
    <Suspense fallback={fallback}>
      <ErrorBoundary fallback={fallback} onError={onError}>
        {children}
      </ErrorBoundary>
    </Suspense>
  );
}
