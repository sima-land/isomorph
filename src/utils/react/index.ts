import { ErrorBoundary, SafeSuspense } from './error-handlers';

export { ErrorBoundary, SafeSuspense };
