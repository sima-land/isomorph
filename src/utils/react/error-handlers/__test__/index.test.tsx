import React from 'react';
import { render } from '@testing-library/react';
import { ErrorBoundary, SafeSuspense } from '..';

describe('ErrorBoundary', () => {
  const error = new Error('Test error');

  const FailedComponent = () => {
    throw error;
  };

  it('should render children component', () => {
    const { container } = render(
      <ErrorBoundary fallback={null}>
        <div>Normal component</div>
      </ErrorBoundary>,
    );

    expect(container.textContent).toContain('Normal component');
    expect(container).toMatchSnapshot();
  });

  it('should render fallback component', () => {
    const { container } = render(
      <ErrorBoundary fallback={<div>Fallback</div>}>
        <>
          <div>Normal component</div>
          <FailedComponent />
        </>
      </ErrorBoundary>,
    );

    expect(container.textContent).toContain('Fallback');
    expect(container).toMatchSnapshot();
  });

  it('should call captureException', () => {
    const handlerException = jest.fn();

    const { container } = render(
      <ErrorBoundary fallback={null} onError={handlerException}>
        <>
          <div>Normal component</div>
          <FailedComponent />
        </>
      </ErrorBoundary>,
    );

    expect(handlerException).toBeCalledWith(error, { componentStack: expect.any(String) });
    expect(container).toMatchSnapshot();
  });
});

describe('SafeSuspense', () => {
  it('should render with props', () => {
    const { container } = render(
      <SafeSuspense fallback={<div>Fallback</div>} onError={jest.fn()}>
        <div>Normal component</div>
      </SafeSuspense>,
    );

    expect(container).toMatchSnapshot();
  });
});
