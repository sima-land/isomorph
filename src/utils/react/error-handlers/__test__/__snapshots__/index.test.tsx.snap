// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary should call captureException 1`] = `<div />`;

exports[`ErrorBoundary should render children component 1`] = `
<div>
  <div>
    Normal component
  </div>
</div>
`;

exports[`ErrorBoundary should render fallback component 1`] = `
<div>
  <div>
    Fallback
  </div>
</div>
`;

exports[`SafeSuspense should render with props 1`] = `
<div>
  <div>
    Normal component
  </div>
</div>
`;
